specs/dynamicdependencies/DynamicDependencies.md

```diff
@@ -0,0 +1,1409 @@
+/// all processes for that user is removed via MddRemovePackageDependency (or
+/// process termination).
+///
+/// MddAddPackageDependency() adds the resolved package to the caller's package graph,
+/// per the rank specified. A process' package graph is a list of packages sorted by
+/// rank in ascending order (-infinityâ€¦0â€¦+infinity). If package(s) are present in the
+/// package graph with the same rank as the call to MddAddPackageDependency the resolved
+/// package is (by default) added after others of the same rank. To add a package
+/// before others o the same rank, specify MddAddPackageDependency::PrependIfRankCollision.
+///
+/// Every MddAddPackageDependency should be balanced by a MddRemovePackageDependency
+/// to remove the entry from the package graph. If the process terminates all package
+/// references are removed, but any pins stay behind.
+///
+/// MddAddPackageDependency adds the resulting package to the process' package
+/// graph, per the rank and options/flags parameters. The process' package
+/// graph is used to search for DLLs (per Dynamic-Link Library Search Order),
+/// WinRT objects and other resources; the caller can now load DLLs, activate
+/// WinRT objects and use other resources from the framework package until
+/// MddRemovePackageDependency is called. The packageDependencyId parameter
+/// must match a package dependency defined for the calling user or the
+/// system (i.e. pinned with MddPinPackageDependency::ScopeIsSystem) else
+/// an error is returned.
+///
+/// @param packageDependencyContext valid until passed to MddRemovePackageDependency()
+/// @param packageFullName allocated via LocalAlloc; use LocalFree to deallocate
+STDAPI MddAddPackageDependency(
+    _In_ PCWSTR packageDependencyId,
+    INT32 rank,
+    MddAddPackageDependency flags,
+    _Out_ MDD_PACKAGEDEPENDENCY_CONTEXT* packageDependencyContext,
+    _Outptr_opt_result_maybenull_ PWSTR* packageFullName);
+
+/// Remove a resolved PackageDependency from the current process' package graph
+/// (i.e. undo MddAddPackageDependency). Used at runtime (i.e. the moral equivalent
+/// of FreeLibrary).
+///
+/// @note This does not unload loaded resources (DLLs etc). After removing
+///        a package dependency any files loaded from the package can continue
+///        to be used; future file resolution will fail to see the removed
+///        package dependency.
+STDAPI_(void) MddRemovePackageDependency(
+    _In_ MDD_PACKAGEDEPENDENCY_CONTEXT packageDependencyContext);
+
+/// Return the package full name that would be used if the
+/// PackageDependency were to be resolved. Does not add the
+/// package to the process graph.
+///
+/// @param packageFullName allocated via LocalAlloc; use LocalFree to deallocate
+STDAPI MddGetResolvedPackageFullNameForPackageDependency(
+    _In_ PCWSTR packageDependencyId,
+    _Outptr_result_maybenull_ PWSTR* packageFullName);
+
+/// Return TRUE if PackageDependency pins would produce the same package
```

Comment:
The comment "would produce the same package when resolved" could be interpreted as saying that this is a valid implementation:

```cpp
// pseudo code
return MddGetResolvedPackageFullNameForPackageDependency(packageDependencyId1) ==
       MddGetResolvedPackageFullNameForPackageDependency(packageDependencyId2);
```

```suggestion
/// Return TRUE if PackageDependency pins would always produce the same package
```

Added "always" to clarify that this is a decision about metaphysical equivalency, and not just "Oh, they happen to resolve to the same package right now (but tomorrow maybe not)".